### PR TITLE
Bundle importer

### PR DESCRIPTION
**Pull request by hatched**
Add the ability to import changeset files by dropping them on the canvas.

### CODE REVIEW

**Comment by hatched** [comment]
#### To QA

File: Makefile [hunk 1]
```diff
@@ -341,7 +341,7 @@ beautify: virtualenv/bin/fixjsstyle
 spritegen: $(SPRITE_GENERATED_FILES)
 
 $(COMPILED_JSX_FILES): $(JSX_FILES)
-	jsx --no-cache-dir -x jsx . .
+	# jsx --no-cache-dir -x jsx . .
```
Comment (hatched) [inline Makefile]: Driveby - this implementation was never finished and causes issues when running tests so it's been commented out until we can finish it.

**Comment by jujugui** [comment]
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/2457/
Test FAILed.

**Comment by jujugui** [comment]
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/2458/
Test FAILed.

File: test/test_bundle_importer.js [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+    });
+
+    it('does not fail if unknown record type', function() {
+      var sortedRecords = [
+        { method: 'badMethod' }
+      ];
+      bundleImporter._executeDryRun(sortedRecords);
+    });
+
+    describe('_execute_addCharm', function() {
```
Comment (makyo) [inline test/test_bundle_importer.js]: These could use an XXX comment with your note about implementing them in the future.

**Comment by makyo** [comment]
:+1: with minor, no QA yet.

File: Makefile (hunk 1 above)
Comment (mitechie) [inline Makefile]: This code isn't going to stay around at all correct? Shouldn't it just get
removed?

Comment (hatched) [inline Makefile]: I felt that just removing the line of code will leave some odd hanging code remaining and I didn't want the branch to creep to much to fully remove the jsx stuff so I elected to comment it out as a interim step.

Comment (mitechie) [inline Makefile]: Can we Chan up as the follow up then please? My understanding was that the
jsx work was in branches not hitting trunk.

**Comment by makyo** [comment]
QA okay, works with old bundles too

**Comment by makyo** [comment]
I should note that this may be a fragile approach to checking the type of the importer file in the future, as SABDFL wants future bundle formats to be yaml as well.  For later, though.

**Comment by hatched** [comment]
@makyo Checking that it's a json file is only for development mode (when no guiserver exists) The typical user will still drop a yaml file which we'll convert behind the scenes.

**Comment by hatched** [comment]
Thanks for the review and qa!

File: app/utils/bundle-importer.js [hunk 3]
```diff
@@ -0,0 +1,426 @@
...
+  }
+
+  BundleImporter.prototype = {
+
+    /**
+      Import a bundle YAML into the current environment.
+
+      @method importBundleYAML
+    */
+    importBundleYAML: function() {},
```
Comment (huwshimi) [inline app/utils/bundle-importer.js]: Is this empty function necessary? I couldn't find anything on this page that uses it.

File: app/utils/bundle-importer.js [hunk 4]
```diff
@@ -0,0 +1,426 @@
...
+      this.recordSet = this._sortDryRunRecords(this.recordSet);
+      this._executeDryRun(this.recordSet);
+    },
+
+    /**
+      Fetch the dry-run output from the Deployer.
+
+      @method fetchDryRun
+    */
+    fetchDryRun: function() {},
```
Comment (huwshimi) [inline app/utils/bundle-importer.js]: Another empty function I couldn't find a reference to.

File: app/utils/bundle-importer.js [hunk 5]
```diff
@@ -0,0 +1,426 @@
...
+                  constraints: {}
+                });
+              }.bind(this, ghostService),
+              // Options used by ECS, ignored by environment.
+              {modelId: ghostService.get('id')});
+          this._saveModelToRequires(record.id, ghostService);
+          next();
+        }.bind(this),
+        'failure': function() {
+          // Create a notification.
```
Comment (huwshimi) [inline app/utils/bundle-importer.js]: Did you actually want to create a notification here?

File: app/utils/bundle-importer.js [hunk 6]
```diff
@@ -0,0 +1,426 @@
...
+      endpoints.forEach(function(ep, index) {
+        endpoints[index][0] = record[ep[0].replace(/^\$/, '')].get('id');
+      }, this);
+      var relationId = 'pending-' + endpoints[0][0] + endpoints[1][0];
+      var relation = this.db.relations.add({
+        relation_id: relationId,
+        'interface': endpoints[0][1].name,
+        endpoints: endpoints,
+        pending: true,
+        scope: 'global', // XXX check the charms to see if this is a subordinate
```
Comment (huwshimi) [inline app/utils/bundle-importer.js]: Is this xxx for something we want to do in the future?

File: test/test_bundle_importer.js [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+          getById: getById
+        }
+      },
+      fakebackend: {
+        _loadCharm: loadCharm
+      }
+    });
+  });
+
+  afterEach(function() {
```
Comment (huwshimi) [inline test/test_bundle_importer.js]: Remove empty afterEach?

**Comment by huwshimi** [comment]
Looks good @hatched! A few tiny notes. No QA. :+1:

File: app/utils/bundle-importer.js (hunk 3 above)
Comment (hatched) [inline app/utils/bundle-importer.js]: This is just a placeholder for when I implement this functionality after hooking up to the backend.

File: app/utils/bundle-importer.js (hunk 4 above)
Comment (hatched) [inline app/utils/bundle-importer.js]: See above.

**Comment by jujugui** [comment]
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/2459/
Test FAILed.

File: app/utils/bundle-importer.js (hunk 6 above)
Comment (hatched) [inline app/utils/bundle-importer.js]: copy pasta error :)

File: test/test_bundle_importer.js (hunk 7 above)
Comment (hatched) [inline test/test_bundle_importer.js]: Unfortunately in order to use the cleanups we need defined before/afterEach methods.

**Comment by jujugui** [comment]
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/2460/
Test FAILed.

**Comment by jujugui** [comment]
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/2461/
Test FAILed.

**Comment by hatched** [comment]
CI is down so manually merging
